components/ScriptLibrary/include/LHVMFile.h

```diff
@@ -0,0 +1,140 @@
+/******************************************************************************
+ * Copyright (c) 2018-2024 openblack developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/openblack/openblack
+ *
+ * openblack is licensed under the GNU General Public License version 3.
+ *******************************************************************************/
+
+#pragma once
+
+#include <cstdint>
+
+#include <array>
+#include <filesystem>
+#include <string>
+#include <vector>
+
+#include "LHVMTypes.h"
+
+namespace openblack::LHVM
+{
+
+struct LHVMHeader
+{
+	std::array<char, 4> magic;
+	LHVMVersion version;
+};
+
+class LHVMFile
+{
+protected:
+	static constexpr const std::array<char, 4> k_Magic = {'L', 'H', 'V', 'M'};
+
+	/// True when a file has been loaded
+	bool _isLoaded {false};
+
+	std::filesystem::path _filename;
+
+	LHVMHeader _header;
+	std::vector<std::string> _variablesNames;
+	std::vector<VMInstruction> _instructions;
+	std::vector<uint32_t> _autostart;
+	std::vector<VMScript> _scripts;
+	std::vector<char> _data;
+	bool _hasStatus {false};
+	VMStack _stack;
+	std::vector<VMVar> _variableValues;
+	std::vector<VMTask> _tasks;
+	uint32_t _ticks {0};
+	uint32_t _currentLineNumber {0};
+	uint32_t _highestTaskId {0};
+	uint32_t _highestScriptId {0};
+	uint32_t _executedInstructions {0};
+
+	/// Error handling
+	void Fail(const std::string& msg);
```

Comment:
Might be asking a bit much because you were shown other components which use `Fail` but since we're moving away from exceptions, would it be possible to design the API without exceptions.

Do discus with us before spending time on this ask.